File: tracks/java/exercises/variables-and-types/ConfigLoader.java
package config;

import java.util.Map;

/**
 * Typed service configuration.
 *
 * Java uses a record here — an immutable data carrier introduced in Java 16.
 * Note: Java primitives (int, double, boolean) have default zero values in
 * class fields, but records require explicit construction.
 */
public record Config(
    String host,
    int port,
    int maxRetries,
    double timeout,
    boolean debug,
    String serviceName
) {

    /** Return a Config with application-level defaults. */
    public static Config defaults() {
        // TODO: implement
        throw new UnsupportedOperationException("not implemented");
    }

    /**
     * Parse raw string key-value pairs into a typed Config.
     * Missing keys should fall back to defaults.
     * Invalid values should throw IllegalArgumentException.
     *
     * Hint: Integer.parseInt, Double.parseDouble, Boolean.parseBoolean
     * Note: Boolean.parseBoolean never throws — it returns false for anything
     * that isn't "true". Is that what you want?
     */
    public static Config load(Map<String, String> raw) {
        // TODO: implement
        throw new UnsupportedOperationException("not implemented");
    }
}

File: tracks/rust/exercises/variables-and-types/config_loader.rs
use std::collections::HashMap;

/// Config holds typed service configuration.
/// Note: Rust has no zero values — every field must be explicitly set.
/// The Default trait is the idiomatic way to handle defaults.
#[derive(Debug, PartialEq)]
pub struct Config {
    pub host: String,
    pub port: i32,
    pub max_retries: i32,
    pub timeout: f64,
    pub debug: bool,
    pub service_name: String,
}

impl Default for Config {
    fn default() -> Self {
        // TODO: return a Config with application-level defaults
        todo!()
    }
}

/// Parse raw string key-value pairs into a typed Config.
/// Missing keys should fall back to defaults.
/// Invalid values should return an Err.
pub fn load_config(raw: &HashMap<String, String>) -> Result<Config, String> {
    // TODO: implement
    // Hint: str::parse::<i32>(), str::parse::<f64>(), str::parse::<bool>()
    Err("not implemented".to_string())
}
